clip/model.py

```diff
@@ -0,0 +1,447 @@
+        self.attention.out_proj.bias = mx.zeros(hidden_dim)
+
+    def init_weights(self, factor: float, num_hidden_layers: int):
+        [embed_dim, _] = self.linear2.weight.shape
+        # Compute std
+        in_proj_std = (embed_dim**-0.5) * ((2 * num_hidden_layers) ** -0.5) * factor
+        out_proj_std = (embed_dim**-0.5) * factor
+        # Initialize attention
+        self.attention.query_proj.weight = normal(
+            shape=self.attention.query_proj.weight.shape, std=in_proj_std
+        )
+        self.attention.key_proj.weight = normal(
+            shape=self.attention.key_proj.weight.shape, std=in_proj_std
+        )
+        self.attention.value_proj.weight = normal(
+            shape=self.attention.value_proj.weight.shape, std=in_proj_std
+        )
+        self.attention.out_proj.weight = normal(
+            shape=self.attention.out_proj.weight.shape, std=out_proj_std
+        )
+        # Initialize MLPs
+        fc_std = (2 * embed_dim) ** -0.5 * factor
+        self.linear1.weight = normal(shape=self.linear1.weight.shape, std=fc_std)
+        self.linear2.weight = normal(shape=self.linear2.weight.shape, std=in_proj_std)
+        self.linear1.bias = mx.zeros_like(self.linear1.bias)
+        self.linear2.bias = mx.zeros_like(self.linear2.bias)
+        # Initialize layer norms
+        self.ln1.bias = mx.zeros_like(self.ln1.bias)
+        self.ln1.weight = mx.ones_like(self.ln1.weight)
+        self.ln2.bias = mx.zeros_like(self.ln2.bias)
+        self.ln2.weight = mx.ones_like(self.ln2.weight)
+
+
+class CLIPTextModel(nn.Module):
+    """Implements the text encoder transformer from CLIP."""
+
+    def __init__(self, config: CLIPTextConfig):
+        super().__init__()
+
+        self.token_embedding = nn.Embedding(config.vocab_size, config.hidden_size)
+        self.position_embedding = mx.zeros(
+            (config.max_position_embeddings, config.hidden_size)
+        )
+        self.layers = [
+            CLIPEncoderLayer(
+                config.hidden_size, config.intermediate_size, config.num_attention_heads
+            )
+            for _ in range(config.num_hidden_layers)
+        ]
+        self.final_layer_norm = nn.LayerNorm(config.hidden_size)
+        self.init_weights(config.initializer_factor)
+
+    def _embed(self, x: mx.array) -> mx.array:
+        # Extract some shapes
+        _, N = x.shape
+        # Compute the embeddings
+        embeddings = self.token_embedding(x)
+        embeddings += self.position_embedding[:N]
+        return embeddings
+
+    def __call__(self, x: mx.array) -> CLIPTextOutput:
+        # Extract some shapes
+        B, N = x.shape
+        eot_tokens = mx.argmax(x, axis=-1)
+        # Look up embeddings
+        x = self._embed(x)
+        # Compute the causal mask
+        mask = nn.MultiHeadAttention.create_additive_causal_mask(N, x.dtype)
+        # Push through the transformer
+        x = reduce(lambda x, l: l(x, mask), self.layers, x)
```

Comment:
Love it :-) . However, I will have to be a bit pedantic and annoying and ask you to change it to

```python
for l in self.layers:
    x = l(x, mask)
```

it is a bit simpler when one needs to add a breakpoint or save some intermediate value or sth. In this particular case it is also not that much more verbose.

Reply:
I agree. The `loop` solution is much better for debugging :)